core/state/trie_prefetcher_extra_test.go

```diff
@@ -0,0 +1,200 @@
+			CleanCacheSize: 3 * 1024 * 1024 * 1024,
+		},
+	}
+	db := NewDatabaseWithConfig(levelDB, tdbConfig)
+	snaps := snapshot.NewTestTree(levelDB, fakeHash(block), root)
+	for count < uint64(wantKVs) {
+		previous := root
+		_ = addBlock(db, snaps, 100_000, 0) // Note this updates root and count
+		b.Logf("Root: %v, kvs: %d, block: %d", root, count, block)
+
+		// Commit every 10 blocks or on the last iteration
+		if block%10 == 0 || count >= uint64(wantKVs) {
+			commit(levelDB, snaps, db)
+			b.Logf("Root: %v, kvs: %d, block: %d (committed)", root, count, block)
+		}
+		if previous != root {
+			require.NoError(db.TrieDB().Dereference(previous))
+		} else {
+			panic("root did not change")
+		}
+	}
+	require.NoError(levelDB.Close())
+	b.Logf("Starting benchmarks")
+	b.Logf("Root: %v, kvs: %d, block: %d", root, count, block)
+	for _, updates := range []int{100, 200, 500} {
+		for _, prefetchers := range []int{0, 1, 4, 16} {
+			b.Run(fmt.Sprintf("updates_%d_prefetchers_%d", updates, prefetchers), func(b *testing.B) {
+				startRoot, startBlock, startCount := root, block, count
+				defer func() { root, block, count = startRoot, startBlock, startCount }()
+
+				levelDB, err := rawdb.NewLevelDBDatabase(path.Join(dir, "level.db"), 0, 0, "", false)
+				require.NoError(err)
+				snaps := snapshot.NewTestTree(levelDB, fakeHash(block), root)
+				db := NewDatabaseWithConfig(levelDB, tdbConfig)
+				if prefetchers > 0 {
+					db = WithPrefetcher(db, prefetchers)
+				}
+				getMetric := func(metric string) int64 {
+					meter := metrics.GetOrRegisterMeter(triePrefetchMetricsPrefix+namespace+"/storage/"+metric, nil)
+					return meter.Snapshot().Count()
+				}
+				startLoads := getMetric("load")
+				for i := 0; i < b.N; i++ {
+					_ = addBlock(db, snaps, updates, prefetchers)
+				}
+				require.NoError(levelDB.Close())
+				b.ReportMetric(float64(getMetric("load")-startLoads)/float64(b.N), "loads")
+			})
+		}
+	}
+}
+
+func fakeHash(block uint64) common.Hash {
+	return common.BytesToHash(binary.BigEndian.AppendUint64(nil, block))
+}
+
+// addKVs adds count random key-value pairs to the state trie of address1 and
+// address2 (each count/2) and returns the new state db and root.
+func addKVs(
+	db Database, snaps *snapshot.Tree,
+	address1, address2 common.Address, root common.Hash, block uint64,
+	count int, prefetchers int,
+) (*StateDB, common.Hash, error) {
+	snap := snaps.Snapshot(root)
+	if snap == nil {
+		return nil, common.Hash{}, fmt.Errorf("snapshot not found")
+	}
+	statedb, err := NewWithSnapshot(root, db, snap)
+	if err != nil {
+		return nil, common.Hash{}, err
+	}
+	if prefetchers > 0 {
+		statedb.StartPrefetcher(namespace)
+		defer statedb.StopPrefetcher()
+	}
+	statedb.SetNonce(address1, 1)
+	for i := 0; i < count/2; i++ {
+		key := make([]byte, 32)
+		value := make([]byte, 32)
+		rand.Read(key)
+		rand.Read(value)
+
+		statedb.SetState(address1, common.BytesToHash(key), common.BytesToHash(value))
+	}
+	statedb.SetNonce(address2, 1)
```

Comment:
This is identical to the `address1` case so they can be collapsed into a `for _, addr := range []...{address1, address1} {` loop.